src/reporters/ScreenshotDiffReporter.ts

```diff
@@ -0,0 +1,254 @@
+import got, { GotPromise } from 'got';
+import { Reporter, Report } from '../typings/reporter';
+import pixelmatch from 'pixelmatch';
+import bluebird from 'bluebird';
+import { S3 } from 'aws-sdk';
+import { GlobalConfig } from '../config';
+
+/* eslint-disable camelcase */
+type Screenshot = {
+  id: string;
+  os: string;
+  os_version: string;
+  browser: string;
+  browser_version: string;
+  url: string;
+  orientation: 'portrait' | 'landscape' | null;
+} & (
+  | {
+      state: 'done';
+      thumb_url: string;
+      image_url: string;
+      created_at: string;
+    }
+  | {
+      state: 'processing' | 'timed-out';
+    });
+
+type PostScreenshotResponse = {
+  job_id: string;
+  state: 'done' | 'queued_all';
+  callback_url: string | null;
+  win_res: string;
+  mac_res: string;
+  quality: 'compressed' | 'original';
+  wait_time: number;
+  orientation: string;
+  screenshots: Screenshot[];
+};
+
+type GetJobResponse = {
+  id: string;
+} & {
+  [K in Exclude<
+    keyof PostScreenshotResponse,
+    'job_id'
+  >]: PostScreenshotResponse[K]
+};
+
+type Browser = {
+  os: string;
+  os_version: string;
+  browser: string;
+  device?: string | null;
+  browser_version?: string | null;
+  real_mobile?: boolean | null;
+};
+
+/* eslint-enable camelcase */
+
+export class ScreenshotDiffReporter implements Reporter {
+  private static API_ENDPOINT = 'https://www.browserstack.com/screenshots';
+
+  private url: string;
+  private bucketName: string;
+  private username: string;
+  private token: string;
+  private s3 = new S3();
+
+  constructor(url: string, config?: Pick<GlobalConfig, 'browserStack'>) {
+    this.url = url;
+
+    if (
+      !config ||
+      !config.browserStack.token ||
+      !config.browserStack.username ||
+      !config.browserStack.screenshotsBucket
+    ) {
+      throw new TypeError(
+        'Invalid configuration provided to ScreenshotDiffReporter constructor',
+      );
+    }
+
+    this.bucketName = config.browserStack.screenshotsBucket;
+    this.token = config.browserStack.token;
+    this.username = config.browserStack.username;
+  }
+
+  private static startScreenshotsJob = async ({
+    url,
+    username,
+    token,
+    browsers,
+  }: {
+    url: string;
+    username: string;
+    token: string;
+    browsers: Browser[];
+  }): Promise<string> => {
+    const { body } = await got.post(ScreenshotDiffReporter.API_ENDPOINT, {
+      auth: `${username}:${token}`,
+      body: {
+        url,
+        browsers,
+      },
+      json: true,
+    });
+
+    return (body as PostScreenshotResponse).job_id;
+  };
+
+  private static getScreenshotsJob = async ({
+    jobId,
+    username,
+    token,
+  }: {
+    jobId: string;
+    username: string;
+    token: string;
+  }): Promise<GotPromise<GetJobResponse>> =>
+    got(`${ScreenshotDiffReporter.API_ENDPOINT}/${jobId}.json`, {
+      auth: `${username}:${token}`,
+      json: true,
+    });
+
+  private static waitForScreenshots = async ({
```

Comment:
Just curious, what's the benefit of making these methods static?

Reply:
Fot this reporter, it does not actually matter. It's only `static` for consistency with the other reporters where it's important to have `static` throttled functions so that the API calls are throttled across _all_ the instances of that reporter. Otherwise, each reporter instance will create a new throttled function which would not be useful at all to avoid rate limits.